scripts/deleteSnapsets.js

```diff
@@ -0,0 +1,243 @@
+fluid.registerNamespace("gpii.dataLoader");
+fluid.setLogging(fluid.logLevel.INFO)
+
+var dbLoader = gpii.dataLoader;
+dbLoader.couchDbUrl = process.argv[2];
+if (!fluid.isValue(dbLoader.couchDbUrl)) {
+    fluid.log("COUCHDB_URL environment variable must be defined");
+    fluid.log("Usage:  node deleteSnapsets.js $COUCHDB_URL");
+    process.exit(1);
+}
+dbLoader.prefsSafesViewUrl = dbLoader.couchDbUrl + "/_design/views/_view/findSnapsetPrefsSafes";
+dbLoader.gpiiKeysViewUrl = dbLoader.couchDbUrl + "/_design/views/_view/findAllGpiiKeys";
+dbLoader.parsedCouchDbUrl = url.parse(dbLoader.couchDbUrl);
+dbLoader.snapsetPrefsSafes = [];
+dbLoader.gpiiKeys = [];
+
+fluid.log("COUCHDB_URL: '" +
+    dbLoader.parsedCouchDbUrl.protocol + "//" +
+    dbLoader.parsedCouchDbUrl.hostname + ":" +
+    dbLoader.parsedCouchDbUrl.port +
+    dbLoader.parsedCouchDbUrl.pathname +
+"'");
+
+/**
+ * Find the Prefs Safes of type "snapset", mark them to be deleted and add
+ * them to an array of records to remove.
+ * @param {String} responseString - The response from the database query for
+ *                                  retrieving the snapset PrefsSafes records
+ * @return {Array} - Array of snapset Prefs Safes each with their "deleted"
+ *                   field set to true.
+ */
+dbLoader.processSnapsets = function (responseString) {
+    fluid.log("Processing the snapset Prefs Safes records...");
+    dbLoader.snapSets = JSON.parse(responseString);
+    fluid.each(dbLoader.snapSets.rows, function (aSnapset) {
+        aSnapset.value._deleted = true;
+        dbLoader.snapsetPrefsSafes.push(aSnapset.value);
+    });
+    fluid.log("\tSnapset Prefs Safes marked for deletion.");
+    return dbLoader.snapsetPrefsSafes;
+};
+
+/**
+ * Find the GPII Key records that are associated with a snapset PrefsSafe, mark
+ * them for deletion, and add them to array of records to delete.
+ * @param {String} responseString - The response from the database query for
+ *                                  retrieving all the GPII Keys.
+ * @return {Array} - Array of snapset PrefsSafes' GPII Keys with the "deleted"
+ *                   field set to true.
+ */
+dbLoader.processGpiiKeys = function (responseString) {
+    fluid.log("Processing the GPII Keys...");
+    var gpiiKeyRecords = JSON.parse(responseString);
+    dbLoader.gpiiKeys = dbLoader.findAndDeletePrefsSafesGpiiKeys(
+        gpiiKeyRecords, dbLoader.snapsetPrefsSafes
+    );
+    fluid.log("\tGPII Keys associated with snapset Prefs Safes marked for deletion.");
+    return dbLoader.gpiiKeys;
+};
+
+/**
+ * Given all the GPII Keys records in the database, find the ones that reference
+ * a snapset PrefsSafe.  As each GPII Key is found it is marked for
+ * deletion.
+ * @param {Array} gpiiKeyRecords - Array of GPII Key records from the database.
+ * @return {Array} - the values from the gpiiKeyRecords that are snapset GPII Keys.
+ */
+dbLoader.findAndDeletePrefsSafesGpiiKeys = function (gpiiKeyRecords, snapSets) {
```

Comment:
This function marks "_delete" flags on GPII keys but doesn't perform deletion. Could you improve the function name? Thanks.

Reply:
I've renamed it `markPrefsSafesGpiiKeysForDeletion()`